src/tree-sitter-language-mode.js

```diff
@@ -947,75 +947,132 @@ class HighlightIterator {
       targetPosition
     );
-    for (let i = this.iterators.length - 1; i >= 0; i--) {
-      this.iterators[i].seek(
-        targetIndex,
-        containingTags,
-        containingTagStartIndices
-      );
+
+    this.iterators = [];
+    const iterator = this.languageMode.rootLanguageLayer.buildHighlightIterator();
+    if (iterator.seek(targetIndex, containingTags, containingTagStartIndices)) {
+      this.iterators.push(iterator);
+    }
+
+    // Populate the iterators array with all of the iterators whose syntax
+    // trees span the given position.
+    for (const marker of injectionMarkers) {
+      const iterator = marker.languageLayer.buildHighlightIterator();
+      if (
+        iterator.seek(targetIndex, containingTags, containingTagStartIndices)
+      ) {
+        this.iterators.push(iterator);
+      }
     }
-    this.iterators.sort((a, b) => b.getIndex() - a.getIndex());
+
+    // Sort the iterators so that the last one in the array is the earliest
+    // in the document, and represents the current position.
+    this.iterators.sort((a, b) => b.compare(a));
+    this.detectCoveredScope();
+
     return containingTags;
   }
 
   moveToSuccessor() {
-    const lastIndex = this.iterators.length - 1;
-    const leader = this.iterators[lastIndex];
-    leader.moveToSuccessor();
-    const leaderCharIndex = leader.getIndex();
-    let i = lastIndex;
-    while (i > 0 && this.iterators[i - 1].getIndex() < leaderCharIndex) i--;
-    if (i < lastIndex) this.iterators.splice(i, 0, this.iterators.pop());
+    // Advance the earliest layer iterator to its next scope boundary.
+    let leader = last(this.iterators);
+    if (leader.moveToSuccessor()) {
+
+      // Maintain the invariant that earlier layer iterators appear later
+      // in the array.
+      const leaderIndex = this.iterators.length - 1;
+      let i = leaderIndex;
+      while (i > 0 && this.iterators[i - 1].compare(leader) < 0) i--;
+      if (i < leaderIndex) {
+        this.iterators.splice(i, 0, this.iterators.pop());
+      }
+    } else {
+      // If the layer iterator was at the end of its syntax tree, then remove
+      // it from the array.
+      this.iterators.pop();
+    }
+
+    this.detectCoveredScope();
+  }
+
+  // Detect whether or not another more deeply-nested language layer has a
+  // scope boundary at this same position. If so, the current language layer's
+  // scope boundary should not be reported.
+  detectCoveredScope() {
+    const layerCount = this.iterators.length;
+    if (layerCount > 1) {
+      const first = this.iterators[layerCount - 1];
+      const next = this.iterators[layerCount - 2];
+      if (
+        next.offset === first.offset &&
+        next.atEnd === first.atEnd &&
+        next.languageLayer.depth > first.languageLayer.depth
+      ) {
+        this.currentScopeIsCovered = true;
+        return;
+      }
+    }
+    this.currentScopeIsCovered = false;
```

Comment:
Ok, I changed the way that we suppress the scope boundaries from the outer layers. I got rid of the concept of "opaque" language layers, and instead I just detect duplicate scope boundaries individually. After each call to `seek` or `moveToSuccessor`, we check if the leading layer iterator is on top of a more deeply-nested iterator. If so, we hide its scope boundaries.

Reply:
Are you concerned about boundaries from a shallower language mode that don't line up to the same offsets as boundaries in the deeper mode? I guess in the Rust macro scenario, the boundaries in the token tree produced by the outer grammar align with the boundaries in the inner grammar, because the tree produced by the inner grammar is going to align with those tokens. Will there ever be cases where scope boundaries in the outer grammar don't align?

If so, I'm wondering if it makes sense to somehow assign a `depth` on the `HighlightIterator` as a whole that can be incremented when we enter an opaque zone and decremented when it ends. Then you could just ignore boundaries from a language mode that's shallower than the current depth regardless of the offset at which they occur. You could sort the deeper layers first and assign the `depth` whenever you start emitting token boundaries on a deeper layer. As soon as that layer gets popped from the array, you could decrement the depth to the next layer.

I suppose if we're not worried about scope boundaries from shallower layers that don't align with boundaries at the deeper layer, this doesn't matter. Just thought I'd share the thought process in case it's valuable.

Reply:
> Are you concerned about boundaries from a shallower language mode that don't line up to the same offsets as boundaries in the deeper mode?

I'm not totally sure, because I don't have a use case in mind, but I actually think it's better that we allow these types non-aligned scopes from the parent layer to still be rendered.

Even though my initial implementation had this concept of "opaque" layers, which would cover up other layers, I don't really think that concept is needed. And if we ever do need it, we might not want to conflate it with the `includeChildren` option. I had conflated the two at first, but then I realized that all I was really trying to solve was the problem of duplicate scope boundaries from inner and outer layers.

It's also kind of tricky to implement the opaque language layer thing, because technically, even an opaque language layer could have "holes" in it, due to holes in its parent layer 😬. My previous code that didn't really handle that situation.